_posts/2024-12-13-python-packaging-security.md

```diff
@@ -0,0 +1,317 @@
+
+```yaml
+  publish:
+    name: >-
+      Publish Python 🐍 distribution 📦 to PyPI
+    if: github.repository_owner == 'pyopensci'
+    needs:
+      - build
+    runs-on: ubuntu-latest
+    environment:
+      name: pypi
+      url: https://pypi.org/p/pyosmeta
+```
+*****
+
+To lockdown a GitHub environment:
+
+* First, go to the <kbd>Settings</kbd> in your repository where the workflow is run
+* Within settings, select **environments** from the left-hand sidebar
+* Add a new environment. Use <kbd>pypi</kbd> as your environment name; this is what PyPA (the Python Packaging Authority) recommends.
+* Ensure <kbd>Required reviewers</kbd> is enabled. This setting allows you to designate specific individuals who can approve and manually run the workflow on GitHub. Any reviewers you add must have the appropriate permissions to authorize the workflow by clicking a button. This adds a human verification step to the process.
+* Once the <kbd>Required reviewers</kbd> button is checked, add maintainers who you want to be able to enable the action to run.
+
+*Optionally, you can click <kbd>prevent self-review</kbd>, preventing someone from triggering a release or a build and then running it!*
+
+<figure>
+    <img src="/images/python-packaging/create-github-environment.gif" alt="Animated gif file that shows the github interface where you can click on settings and go to the environment setting to create or edit a GitHub environment">
+  <figcaption>
+    To create a new environment to use in a GitHub action, 1) go to your repo's settings; 2) click <kbd>environment</kbd>; 3) add a new environment. In this screenshot, we already have a <kbd>pypi</kbd> environment created. Note that you can name your environment whatever you want, however, PyPI suggests that you use the name <kbd>pypi</kbd> for a Trusted Publisher workflow.
+  </figcaption>
+</figure>
+
+
+<figure>
+    <img src="/images/python-packaging/github-action-environment-pypi.png" alt="Screenshot of the GitHub settings interface showing the ‘Environments’ section with configuration options for ‘pypi.’ The ‘Deployment protection rules’ section is visible, with ‘Required reviewers’ enabled and two reviewers listed: ‘lwasser’ and ‘willingc.’ Other options such as ‘Prevent self-review’ and ‘Wait timer’ are present but not enabled.">
+  <figcaption>
+    GitHub environment settings for “pypi,” displaying deployment protection rules with required reviewers configured for workflow approvals.
+  </figcaption>
+</figure>
+
+
+## 2. 🔑 Use Trusted Publisher for PyPI
+
+Now that you have a GitHub environment setup, you can set up Trusted Publisher in your PyPI account.
+
+A Trusted Publisher setup creates a short-lived secure link between PyPI and your repository.
```

Comment:
Maybe remove "short-lived" from this sentence? The underlying implementation of Trusted Publishers uses short-lived credentials, but the link itself is durable!